### PR TITLE
Deepspeech + Conformer corrections

### PR DESCRIPTION
**Pull request by chandramouli-sastry**
This PR contains model code for Deepspeech and a couple small fixes to Conformer pytorch workload. 
Even though the submission.py has batch_size=256, I have not been able to successfully train the pytorch model with this batch-size as it ends up with a cuda out-of-memory error. 
Most notably, the memory utilization increases from 0.88GB to 7GB just after applying the subsampling module and I directed my efforts in fixing this, but did not amount to anything. 
Nonetheless, I also found that when running on a single GPU, I was able to comfortably use a batch size of 32 (i.e., 256/8)!

### CODE REVIEW

**Review by znado — APPROVED** [review]
LGTM!

File: algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py [hunk 1]
```diff
@@ -76,8 +76,8 @@ def preprocess(example):
 
   ds = ds.batch(global_batch_size, drop_remainder=is_training)
 
-  if is_training:
-    ds = ds.repeat()
+  # if is_training:
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/input_pipeline.py]: I have restored this to the earlier version because we have a separate pytorch dataset class now!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 2]
```diff
@@ -239,18 +243,20 @@ def _eval_model_on_split(self,
...
-          targets=batch['targets'].detach().cpu().numpy(),
-          target_paddings=batch['target_paddings'].detach(
-          ).cpu().numpy(),
-          tokenizer=self.tokenizer)
-      loss = self.compute_loss(logits,
-                               logits_padding,
-                               batch['targets'],
-                               batch['target_paddings'])
+
+      with torch.no_grad():
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py]: model_fn also has a torch.no_grad() for the eval pass; this was just for additional guard when doing the greedy-decode. I think this has been removed in the latest commit.

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py [hunk 3]
```diff
@@ -0,0 +1,409 @@
...
+
+    self.encoder_dim = encoder_dim
+    self.input_dropout_rate = input_dropout_rate
+
+    self.conv1 = Conv2dSubsampling(
+        input_channels=1, output_channels=encoder_dim)
+    self.conv2 = Conv2dSubsampling(
+        input_channels=encoder_dim, output_channels=encoder_dim)
+
+    self.lin = nn.LazyLinear(out_features=self.encoder_dim, bias=True)
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py:83]: By doing one forward pass with dummy data in the init_model_fn class.

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py [hunk 4]
```diff
@@ -0,0 +1,409 @@
...
+        output_log_floor=1,
+        pad_end=True,
+        preemph=0.97,
+        preemph_htk_flavor=True,
+        noise_scale=0,
+        num_bins=80,
+        lower_edge_hertz=125,
+        upper_edge_hertz=7600,
+        fft_overdrive=False,
+        output_floor=0.00001)
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py]: Done!

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py [hunk 5]
```diff
@@ -0,0 +1,409 @@
...
+    inputs = self.lin(inputs)
+    inputs = inputs * padding_mask
+    inputs = self.dropout(inputs)
+
+    return inputs
+
+
+class CustomBatchNorm1d(nn.BatchNorm1d):
+
+  def __init__(self, dim, momentum, eps):
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py]: Done!

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py [hunk 6]
```diff
@@ -0,0 +1,409 @@
...
+      in_ = inputs
+    outputs = F.conv2d(
+        input=in_,
+        weight=self.kernel,
+        bias=self.bias,
+        stride=self.filter_stride,
+        dilation=(1, 1),
+        groups=groups)
+
+    outputs = self.bn(outputs.permute(0, 2, 3, 1)).permute(0, 3, 1, 2)
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py:165]: Default BatchNorm in pytorch does not take axis as input, but yeah, permute is cheap in pytorch. If we later find a need to swap this BatchNorm out with a custom batchnorm that track running stats itself (similar to what we did in the conformer class), we can then modify it to take axis as an argument and do away with the permutes?

File: reference_submissions/librispeech_conformer/librispeech_pytorch/submission.py [hunk 7]
```diff
@@ -89,6 +88,8 @@ def update_params(workload: spec.Workload,
     g['lr'] = get_learning_rate(global_step, hyperparameters)
   torch.nn.utils.clip_grad_norm_(current_model.parameters(), max_norm=grad_clip)
   optimizer_state.step()
+  if global_step % 100 == 0:
+    print(f"{global_step}) {train_ctc_loss.item()}")
```
Comment (znado) [inline reference_submissions/librispeech_conformer/librispeech_pytorch/submission.py:92]: I would delete this for now, we can add back in some more logging later as part of the unified logger utils

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/workload.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class LibriSpeechDeepSpeechWorkload(LibriSpeechConformerWorkload):
+
+  def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
+    torch.random.manual_seed(rng[0])
+    model = DeepspeechEncoderDecoder(DeepspeechConfig())
+    self._model = model
+    self.ctc_loss = torch.nn.CTCLoss(blank=0, reduction='none')
+    # Run model once to initialize lazy layers
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/workload.py:27]: nit: comments should have proper punctuation, so add a `.` at the end

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 9]
```diff
@@ -239,6 +234,14 @@ def greedy_decode(self, logits, logit_paddings):
     padding = (fin_result == 0)
     return fin_result, padding
 
+  def sync_sd(self, params):
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:237]: can we rename this to something like sync_batch_norm or something that says it is syncing only BN?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 10]
```diff
@@ -47,6 +49,7 @@ class LibriSpeechConformerWorkload(workload.BaseLibrispeechWorkload):
 
   def init_model_fn(self, rng: spec.RandomState) -> spec.ModelInitState:
     torch.random.manual_seed(rng[0])
+    torch.backends.cudnn.benchmark = False
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:52]: have you confirmed that this is still necessary to prevent an OOM? if so can you add a comment saying why we need this?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 11]
```diff
@@ -113,71 +115,64 @@ def model_fn(
...
+                                      input_paddings.to(DEVICE))
 
     return (logits, logits_paddings), None
 
+  @property
+  def model_params_types(self):
+    if self._param_types is None:
+      self._param_types = param_utils.pytorch_param_types(self._param_shapes)
+    return self._param_types
+
```
Comment (znado) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:128]: we can delete this now that it is in the base class

**Review by znado — APPROVED** [review]
LGTM pending Sourabh model/dataset review

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/libri_dataset.py [hunk 12]
```diff
@@ -0,0 +1,49 @@
+import csv
+
+from absl import logging
+import numpy as np
+import torch
+
+
+class LibriDataset(torch.utils.data.Dataset):
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/libri_dataset.py]: Done!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/libri_dataset.py [hunk 13]
```diff
@@ -0,0 +1,49 @@
+import csv
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/libri_dataset.py:5]: Done!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/libri_dataset.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+
+class LibriDataset(torch.utils.data.Dataset):
+
+  def __init__(self, split, data_dir):
+    super().__init__()
+    self.data_dir = data_dir
+    splits = split.split("+")
+    ids = []
+    for split in splits:
+      logging.info('loading split = %s', split)
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/libri_dataset.py]: Done!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py [hunk 15]
```diff
@@ -463,92 +457,53 @@ def forward(self, outputs, paddings):
...
+    running_var = torch.zeros(config.encoder_dim)
+    self.register_buffer("running_mean", running_mean)
+    self.register_buffer("running_var", running_var)
+    self.weight = nn.Parameter(torch.zeros(config.encoder_dim))
+    self.bias = nn.Parameter(torch.zeros(config.encoder_dim))
+    self.register_buffer("momentum",
+                         torch.FloatTensor([config.batch_norm_momentum]))
+    self.register_buffer("epsilon",
+                         torch.FloatTensor([config.batch_norm_epsilon]))
+    self.register_buffer("dim", torch.FloatTensor([config.encoder_dim]))
```
[the remark this answers is not in the document]
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py]: Done!

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/workload.py (hunk 8 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/workload.py:27]: Done!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 9 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:232]: Actually, I'm syncing the entire state_dict in this function. Should I change it to only sync the BN stats?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 10 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:53]: Yes, added a comment!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 11 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py]: Done! But, it is actually in the init_model_fn?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py [hunk 16]
```diff
@@ -378,8 +368,10 @@ def forward(self,
       if not why_not_fast_path:
```
Comment (sourabh2k15) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py:368]: can we simplify this logic related to why_not_fast_path ?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 17]
```diff
@@ -113,7 +116,8 @@ def model_fn(
 
     with contexts[mode]():
       inputs, input_paddings = augmented_and_preprocessed_input_batch['inputs']
-      logits, logits_paddings = model(inputs, input_paddings)
+      logits, logits_paddings = model(inputs.to(DEVICE),
```
Comment (sourabh2k15) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:119]: do we need to put inputs and input_paddings to device everytime model_fn is called ?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 9 above)
Comment (sourabh2k15) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:232]: what else is present in state_dict apart from batch norm stats ?

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py [hunk 18]
```diff
@@ -0,0 +1,393 @@
...
```
Comment (sourabh2k15) [inline algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py:165]: remove, this huge BatchNorm was causing issues in JAX deepspeech model 
you can also confirm that it's not used in conformer Conv2dSubsampling layer

**Review by sourabh2k15 — CHANGES_REQUESTED** [review]
looks good, left a few comments for conformer and deepspeech model code
dataset iterator LGTM

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 9 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:232]: All the parameters!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py (hunk 16 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/model.py:368]: That's the code written by pytorch developers and I have left it as is because it will be easier to diff the changes!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 17 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:119]: Yes!

File: algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py (hunk 18 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_deepspeech/librispeech_pytorch/model.py]: Done!

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 11 above)
Comment (znado) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py]: yep, my bad

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 9 above)
Comment (runame) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:234]: Sorry I'm out of the loop, but why do we need to sync anything here? And why don't we use the `torch.nn.SyncBatchNorm.convert_sync_batchnorm` wrapper? Also, is this function taking the mean of the parameters (they should be synced already, so this effectively would do nothing)?

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py [hunk 19]
```diff
@@ -253,7 +253,7 @@ def _eval_model_on_split(self,
     data_rng, model_rng = prng.split(rng, 2)
     if split not in self._eval_iters:
       # These iterators repeat indefinitely.
-      self._eval_iters[split] = itertools.cycle(
+      self._eval_iters[split] = (
           self._build_input_queue(
```
Comment (runame) [inline algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:257]: Nit: shift this to the line above.

Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:257]: Because we have our own BatchNorm implementation that takes padding as input and tracks stats. 
Yes, it does nothing for the other parameters.

File: algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py (hunk 9 above)
Comment (chandramouli-sastry) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:234]: Because we have our own BatchNorm implementation that takes padding as input and tracks stats.
Yes, it does nothing for the other parameters.

Comment (runame) [reply algorithmic_efficiency/workloads/librispeech_conformer/librispeech_pytorch/workload.py:234]: I see, maybe it's better to explicitly exclude the parameters to avoid confusion.

**Comment by znado** [comment]
Merging for now, if Runa's comments need addressing we can make a new fix PR.
